fmralign/piecewise.py

```diff
@@ -0,0 +1,457 @@
+# -*- coding: utf-8 -*-
+""" Test module to adapt quickly fmralign for piecewise srm
+Implementation from fastSRM is taken from H. Richard
+"""
+# Author: T. Bazeille
+# License: simplified BSD
+
+import os
+import warnings
+
+import numpy as np
+import nibabel as nib
+from sklearn.base import clone
+from joblib import delayed, Parallel
+from nilearn.image import concat_imgs, load_img
+from sklearn.model_selection import ShuffleSplit
+from sklearn.base import BaseEstimator, TransformerMixin
+from nilearn.maskers._masker_validation import _check_embedded_nifti_masker
+
+from ._utils import _make_parcellation, _intersect_clustering_mask
+
+
+class Identity(BaseEstimator, TransformerMixin):
+    def fit(self, X, y=None):
+        self.basis_list = [np.eye(X[0].shape[0]) for _ in range(len(X))]
+        self.basis_list = [np.eye(X[0].shape[0]) for _ in range(len(X))]
+        return self
+
+    def transform(self, X, y=None, subjects_indexes=None):
+        if subjects_indexes is None:
+            subjects_indexes = np.arange(len(self.basis_list))
+        return np.array([X[i] for i in range(len(subjects_indexes))])
+
+    def inverse_transform(self, X, subjects_indexes=None):
+        return np.array([X for _ in subjects_indexes])
+
+    def add_subjects(self, X_list, S):
+        self.basis_list = [w for w in self.basis_list] + [
+            np.eye(x.shape[0]) for x in X_list
+        ]
+
+
+def generate_X_is(labels, X_list, masker, verbose):
+    """Generate source and target data X_i and Y_i for each piece i.
+
+    Parameters
+    ----------
+    labels : list of ints (len n_features)
+        Parcellation of features in clusters
+    X_list: list of Niimgs
+        Source data
+    masker: instance of NiftiMasker or MultiNiftiMasker
+        Masker to be used on the data. For more information see:
+        http://nilearn.github.io/manipulating_images/masker_objects.html
+    verbose: integer, optional.
+        Indicate the level of verbosity.
+
+    Yields
+    -------
+    X_i: ndarray
+        Source data for piece i (shape : n_samples, n_features)
+    Y_i: ndarray
+        Target data for piece i (shape : n_samples, n_features)
+
+    """
+
+    masked_X_list = []
+    for X in X_list:
+        masked_X_list.append(masker.transform(X))
+    unique_labels = np.unique(labels)
+
+    for k in range(len(unique_labels)):
+        label = unique_labels[k]
+        i = label == labels
+        if (k + 1) % 25 == 0 and verbose > 0:
+            print("Fitting parcel: " + str(k + 1) + "/" + str(len(unique_labels)))
+        # should return X_i Y_i
+
+        yield [X_[:, i].T for X_ in masked_X_list]
+
+
+def fit_one_piece(piece_X_list, method):
+    """Align source and target data in one piece i, X_i and Y_i, using
+    alignment method and learn transformation to map X to Y.
+
+    Parameters
+    ----------
+    X_i: ndarray
```

Comment:
The docstring does not march the signature here.

Reply:
Looks like this was pulled from the existing `fit_one_piece`, which is maybe another plus for integrating it there. Noting that we'll also need to handle `Identity` differently or we'll have overloaded params.